File: src/main/java/cz/nekola/mcpnaut/demo/cli/ArgumentConverter.java
package cz.nekola.mcpnaut.demo.cli;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.core.type.Argument;
import io.modelcontextprotocol.sdk.CallToolRequest; // Java SDK
import jakarta.inject.Singleton;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.List;

@Singleton
public class ArgumentConverter {

    public Object mcpValue2jvmValue(
            CallToolRequest call, // Java SDK CallToolRequest
            @NonNull Argument<?> parameter) {
        JsonNode element = call.getArguments().get(parameter.getName());
        if (element == null || element.isNull()) {
            return null;
        }
        return convertType(parameter.getType(), element);
    }

    private Object convertType(
            @NonNull Class<?> type,
            JsonNode element) {
        if (element.isNull()) return null;

        if (type.equals(Integer.TYPE) || type.equals(Integer.class)) {
            return element.isInt() ? element.asInt() : null;
        } else if (type.equals(Long.TYPE) || type.equals(Long.class)) {
            return element.isLong() || element.isInt() ? element.asLong() : null;
        } else if (type.equals(Boolean.TYPE) || type.equals(Boolean.class)) {
            return element.isBoolean() ? element.asBoolean() : null;
        } else if (type.equals(Float.TYPE) || type.equals(Float.class)) {
            return element.isFloat() || element.isDouble() ? (float) element.asDouble() : null;
        } else if (type.equals(Double.TYPE) || type.equals(Double.class)) {
            return element.isDouble() || element.isFloat() || element.isInt() || element.isLong() ? element.asDouble() : null;
        } else if (type.equals(String.class)) {
            return element.isTextual() ? element.asText() : null;
        } else if (type.isArray()) {
            return convertArray(type, element);
        } else {
            throw new IllegalArgumentException("Unsupported type: " + type);
        }
    }

    private Object convertArray(
            @NonNull Class<?> type,
            JsonNode element) {
        if (!element.isArray()) {
            return null; // Or throw exception, depending on desired strictness
        }
        ArrayNode arrayNode = (ArrayNode) element;
        Class<?> componentType = type.getComponentType();
        int size = arrayNode.size();

        if (componentType.isPrimitive()) {
            if (componentType.equals(Integer.TYPE)) {
                int[] targetArray = new int[size];
                for (int i = 0; i < size; i++) {
                    JsonNode item = arrayNode.get(i);
                    targetArray[i] = (item != null && item.isInt()) ? item.asInt() : 0; // Default to 0 if null/not int
                }
                return targetArray;
            } else if (componentType.equals(Long.TYPE)) {
                long[] targetArray = new long[size];
                for (int i = 0; i < size; i++) {
                    JsonNode item = arrayNode.get(i);
                    targetArray[i] = (item != null && (item.isLong() || item.isInt())) ? item.asLong() : 0L;
                }
                return targetArray;
            } else if (componentType.equals(Double.TYPE)) {
                double[] targetArray = new double[size];
                for (int i = 0; i < size; i++) {
                    JsonNode item = arrayNode.get(i);
                    targetArray[i] = (item != null && item.isNumber()) ? item.asDouble() : 0.0;
                }
                return targetArray;
            } else if (componentType.equals(Float.TYPE)) {
                float[] targetArray = new float[size];
                for (int i = 0; i < size; i++) {
                    JsonNode item = arrayNode.get(i);
                    targetArray[i] = (item != null && item.isNumber()) ? (float) item.asDouble() : 0.0f;
                }
                return targetArray;
            } else if (componentType.equals(Boolean.TYPE)) {
                boolean[] targetArray = new boolean[size];
                for (int i = 0; i < size; i++) {
                    JsonNode item = arrayNode.get(i);
                    targetArray[i] = (item != null && item.isBoolean()) ? item.asBoolean() : false;
                }
                return targetArray;
            } else {
                throw new IllegalArgumentException("Unsupported primitive componentType: " + componentType);
            }
        } else if (componentType.isRecord()) {
            // TODO: Handle record types if necessary
            throw new UnsupportedOperationException("Record types in arrays are not yet supported.");
        } else {
            // Object array
            Object[] targetArray = (Object[]) Array.newInstance(componentType, size);
            for (int i = 0; i < size; i++) {
                targetArray[i] = convertType(componentType, arrayNode.get(i));
            }
            return targetArray;
        }
    }
}

File: src/main/kotlin/cz/nekola/mcpnaut/demo/cli/ArgumentConverter.kt
package cz.nekola.mcpnaut.demo.cli

import io.micronaut.core.annotation.NonNull
import io.micronaut.core.type.Argument
import io.modelcontextprotocol.kotlin.sdk.CallToolRequest
import jakarta.inject.Singleton
import kotlinx.serialization.json.*

/**
 * Converts JSON arguments from a CallToolRequest to their appropriate Java types.
 */
@Singleton
class ArgumentConverter {
    /**
     * Converts a JSON element from a CallToolRequest to the appropriate Java type based on the parameter type.
     *
     * @param call The CallToolRequest containing the arguments as JSON elements
     * @param parameter The parameter information including its type
     * @return The converted value, or null if the conversion is not possible
     * @throws IllegalArgumentException if the parameter type is not supported
     */
    fun mcpValue2jvmValue(
        call: CallToolRequest,
        parameter: @NonNull Argument<*>,
    ): Any? {
        val element: JsonElement = call.arguments[parameter.name]!!
        return convertType(parameter.type, element)
    }

    private fun convertType(
        type: @NonNull Class<*>,
        element: JsonElement,
    ): Any? = when (type) {
        Int::class.javaPrimitiveType,
        Int::class.java,
        Integer::class.java,
        Int::class.javaObjectType,
            -> element.jsonPrimitive.intOrNull

        Long::class.javaPrimitiveType,
        Long::class.java,
        Long::class.javaObjectType,
            -> element.jsonPrimitive.longOrNull

        Boolean::class.javaPrimitiveType,
        Boolean::class.java,
        Boolean::class.javaObjectType,
            -> element.jsonPrimitive.booleanOrNull

        Float::class.javaPrimitiveType,
        Float::class.java,
        Float::class.javaObjectType,
            -> element.jsonPrimitive.floatOrNull

        Double::class.javaPrimitiveType,
        Double::class.java,
        Double::class.javaObjectType,
            -> element.jsonPrimitive.doubleOrNull

        String::class.java,
            -> element.jsonPrimitive.contentOrNull

        else ->
            when {
                type.isArray -> convertArray(type, element)
                else -> throw IllegalArgumentException("Unsupported type $type")
            }
    }

    private fun convertArray(
        type: @NonNull Class<*>,
        element: JsonElement,
    ): Any? {
        val componentType = type.componentType
        return if (componentType.isPrimitive) {
            when (componentType) {
                Int::class.javaPrimitiveType -> {
                    val targetArray = IntArray(element.jsonArray.size)
                    targetArray.indices.forEach { i -> targetArray[i] = element.jsonArray[i].jsonPrimitive.int}
                    targetArray
                }
                Long::class.javaPrimitiveType -> {
                    val targetArray = LongArray(element.jsonArray.size)
                    targetArray.indices.forEach { i -> targetArray[i] = element.jsonArray[i].jsonPrimitive.long}
                    targetArray
                }
                Double::class.javaPrimitiveType -> {
                    val targetArray = DoubleArray(element.jsonArray.size)
                    targetArray.indices.forEach { i -> targetArray[i] = element.jsonArray[i].jsonPrimitive.double}
                    targetArray
                }
                Float::class.javaPrimitiveType -> {
                    val targetArray = FloatArray(element.jsonArray.size)
                    targetArray.indices.forEach { i -> targetArray[i] = element.jsonArray[i].jsonPrimitive.float}
                    targetArray
                }
                Boolean::class.javaPrimitiveType -> {
                    val targetArray = BooleanArray(element.jsonArray.size)
                    targetArray.indices.forEach { i -> targetArray[i] = element.jsonArray[i].jsonPrimitive.boolean}
                    targetArray
                }
                else -> throw IllegalArgumentException("Unsupported componentType: $componentType")
            }
        } else if (componentType.isRecord) {
            TODO()
        } else {
            val elements: List<Any?> = element.jsonArray.map { convertType(componentType, it) }
            val targetArray = java.lang.reflect.Array.newInstance(componentType, elements.size) as Array<Any?>
            targetArray.indices.forEach { i -> targetArray[i] = elements[i] }
            targetArray
        }
    }
}
